lib/Monads/No_Fail.thy

```diff
@@ -225,4 +225,12 @@ lemma no_fail_condition:
   unfolding condition_def no_fail_def
   by clarsimp
 
+lemma no_fail_ex_pre:
+  "(\<And>x. no_fail (P x) f) \<Longrightarrow> no_fail (\<lambda>s. \<exists>x. P x s) f"
+  by (clarsimp simp: no_fail_def)
+
+lemma no_fail_grab_asm:
+  "(G \<Longrightarrow> no_fail P f) \<Longrightarrow> no_fail (\<lambda>s. G \<and> P s) f"
```

Comment:
It's not a big deal, but with this seeming to be the first time one of these lemmas is being used for `no_fail`, is it worth trying to settle on a preferred form of them?
We've never been very consistent but I feel like the `gen_asm` variants are the most common pattern, with the stateless predicate at the end of the precondition and marked by a `K`.

Reply:
The new rules I have for `no_fail` have the exact format and naming approach of `hoare_grab_asm` and `hoare_ex_pre`. So do let me know if I need to change those as well

Reply:
Oh, to clarify, absolutely no problems with this lemma as it is, as you say it completely follows precedent. It just reminded me about how inconsistent we have been around `grab_asm` vs `gen_asm` vs `gen_asm'` vs `gen_asm2` vs ... and I thought that this could be an opportunity to take a more considered approach.

Reply:
Yeah for sure. I'll probably be doing lots of things to lib as a result of the release queue work, so now might be a good time to change those original rules as well

Reply:
I know grab_asm takes from the right side in ccorres, and `wp` also seems to prefer the right, but I don't remember what the corres standard is.

Reply:
The other thing to remember is that `and` (`pred_conj`) is `infixl` while `∧` is `infixr`. This means when you stick to the `and K P` way of doing things, you grab from the right, but if you do `∧ P s` this won't work, meaning you have to grab from the left. Caution is advised in standardisation efforts for this one, although more consistency would be welcome.

Reply:
I think the rule is correct as is, it's completely analogous to `hoare_grab_asm` and has the `G` at the left, which works with `/\` (for `and` you'd want it at the right as Raf is pointing out). Both forms will be useful, i.e. I don't think we can eliminate one in favour of the other.